connector/connect/client/jvm/src/main/scala/org/apache/spark/sql/connect/client/Heartbeat.scala

```diff
@@ -0,0 +1,155 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements.  See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License.  You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.spark.sql.connect.client
+
+import java.util.concurrent.atomic.{AtomicInteger, AtomicLong}
+
+import scala.util.control.NonFatal
+
+import org.apache.spark.internal.Logging
+
+
+/**
+ * This class contains an interface for Heartbeat functionality and the dummy implementation which
+ * doesn't do anything.
+ *
+ * The actual implementation is in HeartbeatImpl
+ */
+private[sql] class Heartbeat {
+  def beat[T](fn: => T): T = fn
+
+  def beatIterator[T](fn: => CloseableIterator[T]): CloseableIterator[T] = fn
+
+  def heartbeatLevel: Int = 0
+
+  def pingCount: Long = 0
+}
+
+/**
+ * This class implements Heartbeat. Heartbeat consists of sending a dummy request once in a while
+ * if there is at least one long-term query running.
+ *
+ * Otherwise the connection may disconnect before client receives the response back.
+ */
+private[sql] class HeartbeatImpl(bstub: CustomSparkConnectBlockingStub)
+    extends Heartbeat
+    with Logging {
+  val beatLevel: AtomicInteger = new AtomicInteger(0)
+  var count: AtomicLong = new AtomicLong(0)
+
+  override def heartbeatLevel: Int = beatLevel.get()
+
+  override def pingCount: Long = count.get()
+
+  private val pingThread = new Thread() {
```

Comment:
Why not use an ExecutorService and schedule a runnable?

Reply:
Sure, I will look into it